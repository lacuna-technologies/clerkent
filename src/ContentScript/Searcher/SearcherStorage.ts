import Storage from '../../utils/Storage'

const KEYS = {
  LAWNET_QUERY: `LAWNET_QUERY`,
  LEXIS_UK_QUERY: `SEARCHER_LEXIS_UK_QUERY`,
}

const storeLexisUKQuery = (query: string) => Storage.set(KEYS.LEXIS_UK_QUERY, query)
const getLexisUKQuery = () => Storage.get(KEYS.LEXIS_UK_QUERY)
const removeLexisUKQuery = () => Storage.remove(KEYS.LEXIS_UK_QUERY)

const storeLawNetQuery = (query: string) => Storage.set(KEYS.LAWNET_QUERY, query)
const getLawNetQuery = () => Storage.get(KEYS.LAWNET_QUERY)
const removeLawNetQuery = () => Storage.remove(KEYS.LAWNET_QUERY)

const SearcherStorage = {
  getLawNetQuery,
  getLexisUKQuery,
  removeLawNetQuery,
  removeLexisUKQuery,
  storeLawNetQuery,
  storeLexisUKQuery,
}

export default SearcherStorage