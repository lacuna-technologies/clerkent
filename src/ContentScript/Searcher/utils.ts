export const createInput = (key: string, value: string) => {
  const newInput = document.createElement(`input`)
  newInput.setAttribute(`type`,`hidden`)
  newInput.setAttribute(`name`,key)
  newInput.setAttribute(`value`, value)
  return newInput
}

export const postFormData = (url: string, data: Record<string, string | number>[]) => {
  const form = document.createElement(`form`)
  form.setAttribute(`method`,`post`)
  form.setAttribute(`action`, url)

  for (const attribute of data) {
    const [key] = Object.keys(attribute)
    const [value] = Object.values(attribute)
    const newInput = createInput(key, `${value}`)
    form.append(newInput)
  }

  document.querySelectorAll(`body`)[0].append(form)
  form.submit()
}