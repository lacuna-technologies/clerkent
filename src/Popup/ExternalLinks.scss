@import "../styles/core.scss";

#external-links {

  > .links-container {
    display: flex;
    flex-direction: row;
    justify-content: center;
    margin-top: 0.5rem;

    > a {
      @include no-link;
      padding: 0.2rem 0.6rem;
      border: 1px solid $border-color;
      border-radius: 0.25rem;

      &:hover {
        background-color: lightgray;
      }

      & + a {
        margin-left: 1rem;
      }
    }
  }
  
}