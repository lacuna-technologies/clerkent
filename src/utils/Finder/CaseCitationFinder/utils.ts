import Logger from "utils/Logger"

export const formatAbbr = (
  { abbr, appendum = null }:
  { abbr: string, appendum?: string },
) => {
  const end = appendum || ``
  return `${abbr
    .split(``)
    .map((letter: string) =>
      /[a-z]/i.test(letter)
        ? `${letter}\\.?`
        : letter,
    ).join(``)
    }${end}`
}

export const formatAbbrs = (
  abbrArray: { abbr: string, appendum?: string }[],
) => abbrArray.map(
  ({ abbr, appendum }) => formatAbbr({ abbr, appendum}),
).join(`|`)

export const cleanVolume = (volumeString: string) => volumeString.replace(/\./g, ``).toLowerCase().trim()

export const sortCitationsByVolume = (
  abbrsList: { abbr: string, appendum?: string }[],
  citationsArray: string[],
): string[] => {
  const lastIfNotFound = (index: number) => (
    index === -1
      ? citationsArray.length + 1
      : index
  )
  return citationsArray.sort((a, b) => {
    const indexA = lastIfNotFound(
      abbrsList.findIndex(
        currentAbbr => new RegExp(
          formatAbbr(currentAbbr),
          `i`,
        ).test(a),
      ),
    )
    const indexB = lastIfNotFound(
      abbrsList.findIndex(currentAbbr => 
        new RegExp(formatAbbr(currentAbbr), `i`).test(b),
      ),
    )
    return indexA - indexB
  })
}

export const sortCasesByVolume = (
  abbrsList: { abbr: string, appendum?: string }[],
  citationsArray: Law.Case[],
  attribute = `citation`,
): Law.Case[] => sortCitationsByVolume(
    abbrsList,
    (citationsArray).map(c => c[attribute]),
  ).map(c => (citationsArray).find(v => v[attribute] === c))