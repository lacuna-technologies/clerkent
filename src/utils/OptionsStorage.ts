import Storage from './Storage'

export type InstitutionalLogin = `KCL` | `LSE` | `UCL` | `NTU` | `NUS` |  `SMU` | `None`

const keysObject = {
  OPTIONS_CLIPBOARD_PASTE_ENABLED: {
    defaultValue: false as boolean,
    shortName: `clipboardPaste`,
  },
  OPTIONS_HIGHLIGHT_ENABLED: {
    defaultValue: true as boolean,
    shortName: `highlight`,
  },
  OPTIONS_INSTITUTIONAL_LOGIN: {
    defaultValue: `None` as InstitutionalLogin,
    shortName: `institutionalLogin`,
  },
} as const
export type OptionFullKey = keyof typeof keysObject
export type OptionShortName = typeof keysObject[OptionFullKey][`shortName`]
export type OptionType<T extends OptionFullKey> = typeof keysObject[T][`defaultValue`]
export type OptionsSettings = {
  [K in OptionFullKey]: OptionType<K>
}

const defaultOptions = Object.entries(keysObject).reduce((accumulator, [key, { defaultValue }]) => ({
  ...accumulator,
  [key]: defaultValue,
}), {} as OptionsSettings)

const makeGet = <K extends OptionFullKey>(settingKey: K) => async (): Promise<OptionsSettings[K]> => {
  const result = await Storage.get(settingKey)
  if(result === null){
    return defaultOptions[settingKey]
  }
  return result
}

const makeSet = <K extends OptionFullKey>(settingKey: K) => (
  value: typeof keysObject[K][`defaultValue`],
) => Storage.set(
  settingKey,
  value,
)

export type OptionStorageContentType = {
  [Property in OptionFullKey as typeof keysObject[OptionFullKey][`shortName`]]: {
    get: () => Promise<OptionType<Property>>,
    set: (value: OptionType<Property>) => void,
  }
}

const optionStorageContent = Object.entries(keysObject).reduce((
  accumulator,
  [key, { shortName }] : [OptionFullKey, typeof keysObject[OptionFullKey]]) => ({
  ...accumulator,
  [shortName]: {
    get: makeGet(key as OptionFullKey),
    set: makeSet(key as OptionFullKey),
  },
}), {} as OptionStorageContentType)

const getAll = async () => {
  const allResults = await Promise.all(
    Object.values(keysObject).map(({ shortName }) => optionStorageContent[shortName].get()),
  )

  return Object.entries(keysObject).reduce((accumulator, [key], index) => ({
    ...accumulator,
    [key]: allResults[index],
  }), {} as OptionsSettings)
}

const OptionsStorage = {
  ...optionStorageContent,
  defaultOptions,
  getAll,
}

export default OptionsStorage