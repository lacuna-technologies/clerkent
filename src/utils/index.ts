export { default as Constants } from './Constants'
export { default as Finder } from './Finder'
export { default as Messenger } from './Messenger'
export { default as Request } from './Request'
export { default as Storage } from './Storage'
export { default as Logger } from './Logger'
export { default as Helpers } from './Helpers'
export { default as Clipboard } from './Clipboard'