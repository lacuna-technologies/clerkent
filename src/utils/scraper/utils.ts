import Fuse from 'fuse.js'
import Helpers from '../Helpers'
import Constants from 'utils/Constants'
import Storage from 'utils/Storage'
import Finder from "../Finder"

export const findCitation = (function_ = Finder.findCaseCitation, inputText: string) => {
  const results = function_(inputText)
  if(results.length > 0){
    return results[0].citation
  }
  return null
}

const longestCommonSubstring = (stringA: string, stringB: string) => {
  if (!stringA || !stringB) {
    return {
      length: 0,
      str: ``,
    }
  }

  let longestSubstr = ``
  let currentSubstr = ``
  for (const element of stringA) {
    currentSubstr += element
    if (!stringB.toLowerCase().includes(currentSubstr.toLowerCase())) {
      if (currentSubstr.length > longestSubstr.length) {
        longestSubstr = currentSubstr
      }
      currentSubstr = ``
    }
  }
  return currentSubstr.length > longestSubstr.length ? {
    length: currentSubstr.length,
    str: currentSubstr,
  } : {
    length: longestSubstr.length,
    str: longestSubstr,
  }
}

const probablySameCase = (caseNameA: string, caseNameB: string) => {
  const {
    length: longestSubstrLength,
  } = longestCommonSubstring(
    caseNameA,
    caseNameB,
  )
  const sameCaseProbability = longestSubstrLength / Math.min(caseNameA.length, caseNameB.length)
  return sameCaseProbability >= 0.8
}

export const sortByName = (query: string, cases: Law.Case[]) => {
  const cleanQuery = Helpers.removeCommonAppends(query.toLowerCase())
  const fuse = new Fuse(cases.map(({ name }) => Helpers.removeCommonAppends(name.toLowerCase())), { ignoreLocation: true })
  return fuse
    .search(cleanQuery)
    .map(({ refIndex }) => cases[refIndex])
}

export const databaseUse = async (jurisdictionId: Law.JurisdictionCode, databaseId: string, function_: () => Promise<Law.Case[]>): Promise<Law.Case[]> => {
  const databasesStatus: typeof Constants.DEFAULT_DATABASES_STATUS = await Storage.get(`DATABASES_STATUS`) || Constants.DEFAULT_DATABASES_STATUS
  if(databasesStatus[jurisdictionId][databaseId]){
    return function_()
  } 
  return []
}