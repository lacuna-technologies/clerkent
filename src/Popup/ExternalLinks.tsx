import React, { useEffect, useState , useCallback } from 'react'
import type Law from '../types/Law'
import Constants from '../utils/Constants'
import OptionsStorage from '../utils/OptionsStorage'
import type { OptionsSettings, InstitutionalLogin } from '../utils/OptionsStorage'
import './ExternalLinks.scss'

import SearcherStorage from '../ContentScript/Searcher/SearcherStorage'

interface Props {
  jurisdiction: Law.JursidictionCode
  type: Law.Type
  query: string
}

const getLawNetURL = (
  institution: OptionsSettings[`OPTIONS_INSTITUTIONAL_LOGIN`],
  query: string,
) => (new Proxy({
  NTU: `http://www.lawnet.sg.remotexs.ntu.edu.sg/lawnet/group/lawnet/legal-research/basic-search?clerkent-query=${query}`,
  NUS: `https://proxylogin.nus.edu.sg/lawproxy1/public/login.asp?logup=false&url=https://www.lawnet.sg/lawnet/web/lawnet/ip-access?clerkent-query=${query}`,
  SMU: `https://login.libproxy.smu.edu.sg/login?auth=shibboleth&url=https://www.lawnet.sg/lawnet/web/lawnet/ip-access?clerkent-query=${query}`,
}, {
  get(target, property) {
    if(property in target) {return target[property as string]}
    return `https://www.lawnet.sg/?clerkent-query=${query}`
  },
})[institution])

const ExternalLinks: React.FC<Props> = ({
  jurisdiction,
  query,
  type,
}) => {
  const [institution, setInstitution] = useState(OptionsStorage.defaultOptions.OPTIONS_INSTITUTIONAL_LOGIN)

  const onLawnetClick = useCallback(() => {
    // this is necessary because the NUS and SMU proxies don't pass query params
    // and the redirect occurs before content scripts have time to run
    if([`NUS`, `SMU`, `NTU`].includes(institution)){
      SearcherStorage.storeLawNetQuery(query)
    }
  }, [query, institution])

  useEffect(() => {
    (async () => {
      const institutionSelection = await OptionsStorage.institutionalLogin.get() as InstitutionalLogin
      setInstitution(institutionSelection)
    })()
  }, [])

  return (
    <div id="external-links">
      <small>Search on:</small>
      <div className="links-container">
        {
          jurisdiction === Constants.JURISDICTIONS.UK.id ? (
            <>
              <a
                href={`https://uk.westlaw.com/Browse/Home/WestlawUK/Cases?clerkent-query=${query}&clerkent-type=${type}`}
                target="_blank" rel="noreferrer"
              >Westlaw UK</a>

              <a
                href={`https://www.lexisnexis.com/uk/legal/home/home.do?clerkent-query=${query}`}
                target="_blank" rel="noreferrer"
              >LexisNexis UK</a>

              <a
                href={`https://app.justis.com/search/${query}/1/Relevance`}
                target="_blank" rel="noreferrer"
              >Justis</a>
            </>
          ) : null
        }
        {
          jurisdiction === Constants.JURISDICTIONS.SG.id ? (
            <>
              <a
                href={getLawNetURL(institution, query)}
                onClick={onLawnetClick}
                target="_blank" rel="noreferrer"
              >LawNet</a>
            </>
          ) : null
        }
        {
          [
            Constants.JURISDICTIONS.CA.id,
            Constants.JURISDICTIONS.HK.id,
            Constants.JURISDICTIONS.AU.id,
            Constants.JURISDICTIONS.NZ.id,
            Constants.JURISDICTIONS.EU.id,
            Constants.JURISDICTIONS.ECHR.id,
            Constants.JURISDICTIONS.UN.id,
          ].includes(jurisdiction) ? (
            <>
              <a
                href={`https://app.justis.com/search/${query}/1/Relevance`}
                target="_blank" rel="noreferrer"
              >Justis</a>
            </>
          ) : null
        }
      </div>
    </div>
  )
}

export default ExternalLinks