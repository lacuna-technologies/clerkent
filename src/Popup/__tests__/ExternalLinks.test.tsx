import { render } from '@testing-library/preact'
import ExternalLinks from 'Popup/components/ExternalLinks'
import { Constants } from 'utils'
import OptionsStorage from 'utils/OptionsStorage'
import SearcherStorage from 'ContentScript/Searcher/SearcherStorage'

jest.mock(`utils/OptionsStorage`)
const institutionalLoginGet = OptionsStorage.institutionalLogin.get as jest.Mock

jest.mock(`ContentScript/Searcher/SearcherStorage`)
const storeLawNetQuery = SearcherStorage.storeLawNetQuery as jest.Mock

const TYPE: Law.Type = `case-name`
const QUERY = `abc123`

describe(`ExternalLinks`, () => {

  beforeAll(() => {
    institutionalLoginGet.mockImplementation(() => Promise.resolve(`None`))
    storeLawNetQuery.mockImplementation(() => Promise.resolve())
  })

  it(`renders without error for UK`, () => {
    const tree = render(
      <ExternalLinks
        jurisdiction={Constants.JURISDICTIONS.UK.id}
        type={TYPE}
        query={QUERY}
      />,
    )
    expect(tree).toMatchSnapshot()
  })

  it(`renders without error for Singapore`, () => {
    const tree = render(
      <ExternalLinks
        jurisdiction={Constants.JURISDICTIONS.SG.id}
        type={TYPE}
        query={QUERY}
      />,
    )
    expect(tree).toMatchSnapshot()
  })

  it(`renders without error for EU`, () => {
    const tree = render(
      <ExternalLinks
        jurisdiction={Constants.JURISDICTIONS.EU.id}
        type={TYPE}
        query={QUERY}
      />,
    )
    expect(tree).toMatchSnapshot()
  })
})