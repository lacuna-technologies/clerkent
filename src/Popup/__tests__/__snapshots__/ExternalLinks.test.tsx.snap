// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalLinks renders without error for EU 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="select-none mt-4"
      >
        <small>
          Search on:
        </small>
        <div
          class="flex flex-row justify-center mt-2 gap-4"
        >
          <a
            class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
            href="https://google.com/search?hl=en&q=abc123"
            rel="noreferrer"
            target="_blank"
          >
            Google
          </a>
          <a
            class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
            href="https://app.justis.com/search/abc123/1/Relevance"
            rel="noreferrer"
            target="_blank"
          >
            Justis
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="select-none mt-4"
    >
      <small>
        Search on:
      </small>
      <div
        class="flex flex-row justify-center mt-2 gap-4"
      >
        <a
          class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
          href="https://google.com/search?hl=en&q=abc123"
          rel="noreferrer"
          target="_blank"
        >
          Google
        </a>
        <a
          class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
          href="https://app.justis.com/search/abc123/1/Relevance"
          rel="noreferrer"
          target="_blank"
        >
          Justis
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ExternalLinks renders without error for Singapore 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="select-none mt-4"
      >
        <small>
          Search on:
        </small>
        <div
          class="flex flex-row justify-center mt-2 gap-4"
        >
          <a
            class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
            href="https://google.com/search?hl=en&q=abc123"
            rel="noreferrer"
            target="_blank"
          >
            Google
          </a>
          <a
            class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
            href="https://www.lawnet.sg/?clerkent-query=abc123"
            rel="noreferrer"
            target="_blank"
          >
            LawNet
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="select-none mt-4"
    >
      <small>
        Search on:
      </small>
      <div
        class="flex flex-row justify-center mt-2 gap-4"
      >
        <a
          class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
          href="https://google.com/search?hl=en&q=abc123"
          rel="noreferrer"
          target="_blank"
        >
          Google
        </a>
        <a
          class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
          href="https://www.lawnet.sg/?clerkent-query=abc123"
          rel="noreferrer"
          target="_blank"
        >
          LawNet
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ExternalLinks renders without error for UK 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="select-none mt-4"
      >
        <small>
          Search on:
        </small>
        <div
          class="flex flex-row justify-center mt-2 gap-4"
        >
          <a
            class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
            href="https://google.com/search?hl=en&q=abc123"
            rel="noreferrer"
            target="_blank"
          >
            Google
          </a>
          <a
            class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
            href="https://uk.westlaw.com/Browse/Home/WestlawUK/Cases?clerkent-query=abc123&clerkent-type=case-name"
            rel="noreferrer"
            target="_blank"
          >
            Westlaw UK
          </a>
          <a
            class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
            href="https://www.lexisnexis.com/uk/legal/home/home.do?clerkent-query=abc123"
            rel="noreferrer"
            target="_blank"
          >
            LexisNexis UK
          </a>
          <a
            class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
            href="https://app.justis.com/search/abc123/1/Relevance"
            rel="noreferrer"
            target="_blank"
          >
            Justis
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="select-none mt-4"
    >
      <small>
        Search on:
      </small>
      <div
        class="flex flex-row justify-center mt-2 gap-4"
      >
        <a
          class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
          href="https://google.com/search?hl=en&q=abc123"
          rel="noreferrer"
          target="_blank"
        >
          Google
        </a>
        <a
          class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
          href="https://uk.westlaw.com/Browse/Home/WestlawUK/Cases?clerkent-query=abc123&clerkent-type=case-name"
          rel="noreferrer"
          target="_blank"
        >
          Westlaw UK
        </a>
        <a
          class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
          href="https://www.lexisnexis.com/uk/legal/home/home.do?clerkent-query=abc123"
          rel="noreferrer"
          target="_blank"
        >
          LexisNexis UK
        </a>
        <a
          class="py-1 px-2 border border-solid border-gray-400 rounded hover:bg-gray-200"
          href="https://app.justis.com/search/abc123/1/Relevance"
          rel="noreferrer"
          target="_blank"
        >
          Justis
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
